scipy/ndimage/tests/test_interpolation.py

```diff
@@ -51,6 +52,32 @@ def shift(x):
             ndimage.geometric_transform(data, shift, cval=-1, mode=mode,
                                         output_shape=(4,)))
 
+    @pytest.mark.parametrize('mode', ['mirror', 'reflect', 'grid-mirror',
+                                      'grid-wrap'])
+    @pytest.mark.parametrize('order', range(6))
+    def test_boundary_spline_accuracy(self, mode, order):
+        """Tests based on examples from gh-2640"""
+        data = numpy.array([-2, -1, 0, 1, 2], float)
+        x = numpy.linspace(-6, 10, num=1000)
+        y = ndimage.map_coordinates(data, [x], order=order, mode=mode)
+
+        # compute expected value using explicit padding via numpy.pad
+        npad = 100
+        pad_mode = {
+            'mirror': 'reflect',
+            'reflect': 'symmetric',
+            'grid-mirror': 'symmetric',
+            'grid-wrap': 'wrap',
+        }.get(mode)
+        padded = numpy.pad(data, npad, mode=pad_mode)
+        expected = ndimage.map_coordinates(padded, [npad + x], order=order,
+                                           mode=mode)
+
+        if mode in ['reflect', 'grid-mirror'] and order > 1:
+            numpy.testing.assert_allclose(y, expected, rtol=1e-4, atol=1e-13)
```

Comment:
I looked into this relatively high error for mode='reflect' a bit more. It actually is not a problem if the original signal is > 10 samples in length or so. This must be some limitation of the implementation of that boundary mode for short signal lengths.

If we change the test signal above to something like `data = numpy.arange(-3, 8, dtype=float)`, the tests will pass with `rtol=1e-7` in all cases.

Reply:
I think the inaccuracy is introduced here where the summation is only over a small number of elements when the length, `n`, is short. For longer `n`, `z_i = z**n` becomes close to zero.

https://github.com/scipy/scipy/blob/30e9c92eb89cfc92d6988fb77402475c24f90758/scipy/ndimage/src/ni_splines.c#L218-L222

To fix this for short signals, we would have to operate on an extended signal instead of the original signal length.

For long signals, this loop can probably be truncated at something `min(n, 32)` samples. This is because `z_i` becomes extremely over time in the loop (all filter poles, z, are < 1.0).

The largest pole, `z`, for any of the implemented splines is ~0.43 for the 5th order spline.  `0.43 ** 32` = 1.8e-12, so there is not much point in going to many additional terms.

Reply:
I think this is not worth the effort to improve to be honest. Signal that short are uncommon and will have bigger issues than slightly large `rtol` for signal extrapolation.